File: boj/brute-force/1248.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int N;
    static int[][] oper;
    static int[] result;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        result = new int[N];

        int operIdx = 0;
        oper = new int[N][N];
        char[] input = br.readLine().toCharArray();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (i > j) continue;

                if (input[operIdx] == '-') {
                    oper[i][j] = -1;
                }
                if (input[operIdx] == '+') {
                    oper[i][j] = 1;
                }
                if (input[operIdx] == '0') {
                    oper[i][j] = 0;
                }

                operIdx++;
            }
        }

        search(0);
        for (int i = 0; i < N; i++) {
            bw.write(result[i] + " ");
        }
        bw.flush();
    }

    static boolean search(int index) {
        if (index == N) {
            return true;
        }

        if (oper[index][index] == 0) {
            result[index] = 0;
            return isInRange(index) && search(index + 1);
        }

        for (int i = 1; i <= 10; i++) {
            result[index] = oper[index][index] * i;
            if (isInRange(index) && search(index + 1)) return true;
        }

        return false;
    }

    static boolean isInRange(int index) {

        for (int i = 0; i <= index; i++) {
            int sum = 0;
            for (int j = i; j <= index; j++) {
                sum += result[j];
            }
            if (oper[i][index] > 0 && sum <= 0) return false;
            if (oper[i][index] < 0 && sum >= 0) return false;
            if (oper[i][index] == 0 && sum != 0) return false;
        }

        return true;
    }

}

File: boj/brute-force/1248.swift
import Foundation

var N = Int()
var S = [[String]]()
var result = [Int]()

func calc(_ index: Int) -> Bool {
    
    for i in (0...index).reversed() {
        var sum = 0
        for j in i...index {
            sum += result[j]
        }
        
        if S[i][index] == "-" && sum >= 0 { return false }
        if S[i][index] == "+" && sum <= 0 { return false }
        if S[i][index] == "0" && sum != 0 { return false }
    }
    
    return true
}

func search(_ index: Int) {
    if index == N {
        print(result.inline)
        exit(0)
    }

    if S[index][index] == "0" {
        result[index] = 0
        if calc(index) { search(index+1) }
    }

    else if S[index][index] == "-" {
        for num in -10..<0 {
            result[index] = num
            if calc(index) { search(index+1) }
        }
    }
    else { // "+"
        for num in 0...10 {
            result[index] = num
            if calc(index) { search(index+1) }
        }
    }
}


if let input = readLine() {
    if let n = Int(input) {
        N = n
        guard let inputOfS = readLine()?.compactMap({String($0)}) else { exit(0) }
        S = [[String]](repeating: [String](repeating: String(), count: N), count: N)
        result = [Int](repeating: 0, count: N)

        var head = 0
        for i in 0..<N {
            for j in i..<N {
                S[i][j] = inputOfS[head]
                head+=1
            }
        }

        search(0)
    }
}

extension Array {
    var inline: String {
        var str = String()
        for element in self {
            str += "\(element) "
        }
        return str
    }
}
